### ISSUE TITLE
Burn's scatter values backward bug

### ISSUE THREAD

**Issue by khoek** [tracel-ai/burn#4063]
**Describe the bug**
Burn's `x.scatter(d, indices, values)` requires that `indices` and `values` have the same dimensions, but doesn't make the same requirement for `x` (and indeed there is no reason why one would want to have such a requirement, given that entries of `indices` could range over whatever bounds are appropriate for `x`).

However, while `x.scatter(1, indices, values)` works fine when `x` has dimensions different to `indices` and `values` at inference time (as expected), with e.g. Autodiff<B<f32>> for B any of NdArray, Candle, Cuda, or Cpu (but not Wgpu or Tch) the whole stack crashes after a `.backward()` due to an assertion/memory access violation depending on the backend. It happens because, internally, autodiff attempts to use the backend's ordinary `.scatter()` (to propagate gradients backwards), and since `x` and `indices` have different dimensions this immediately blows up.

I'm no expert, but I think the implementation of scatter autodiff is just broken:

https://github.com/tracel-ai/burn/blob/66b51fc5ebb514c96c874ca25ab109cc9d504749/crates/burn-autodiff/src/ops/tensor.rs#L991C1-L1014C10
```rust
            fn backward(
                self,
                ops: Ops<Self::State, 2>,
                grads: &mut Gradients,
                _checkpointer: &mut Checkpointer,
            ) {
                let (dim, indices, shape_lhs, shape_rhs, device) = ops.state;
                let [indices_4lhs, indices_4rhs] = duplicate(&ops.parents, Some(indices));


                binary::<B, _, _>(
                    ops.parents,
                    ops.node,
                    grads,
                    |grad| {
                        let zeros = B::float_zeros(shape_lhs, &device, grad.dtype().into());
                        B::float_scatter(dim, grad, indices_4lhs.unwrap(), zeros)
                    },
                    |grad| {
                        let zeros = B::float_zeros(shape_rhs, &device, grad.dtype().into());
                        B::float_scatter(dim, zeros, indices_4rhs.unwrap(), grad)
                    },
                );
            }
        }
```

I think a correct implementation should simply be something like (note the `rhs` in the first line):
```rust
|grad| grad
|grad| B::float_gather(dim, grad, indices_4rhs.unwrap()),
```

Given "equal size tensors" (avoiding the crash), I think the current implementation still (wrongly) zeroes entries of the tensor gradient direction which are not written, as well as permuting the value gradients according to index order which is also wrong?

Perhaps this happened because the author(s) didn't realize that Burn's scatters have "add" and not "write" semantics?

**Reproducer**
```rust
use burn::backend::{Autodiff, Candle, Cuda, NdArray, LibTorch, Wgpu};
use burn::tensor::backend::AutodiffBackend;
use burn::tensor::{Int, Tensor, TensorData};

fn scatter_notequal_forward_to_backward<B: AutodiffBackend>() {
    let device = B::Device::default();

    let flat_len = 1000000;
    let active_len = 30000;
    let batch = 1;

    let x = Tensor::<B, 2>::zeros([batch, flat_len], &device).require_grad();

    let idx_host: Vec<i32> = (0..active_len as i32).collect();
    let indices_1d = Tensor::<B, 1, Int>::from_ints(idx_host.as_slice(), &device);\
    let indices = indices_1d.reshape([batch, active_len]);

    let values = Tensor::<B, 2>::from_data(
        TensorData::new((0..active_len).map(|i| i as f32).collect::<Vec<f32>>(), [batch, active_len]),
        &device,
    );

    // Always works
    let y = x.clone().scatter(1, indices, values);

    let loss = y.sum();

    // Crashes (depending on backend)
    let grads = loss.backward();

    // Force e.g. Cuda to synchronize and actually trigger the crash
    println!("{}", x.grad(&grads).unwrap().sum().to_data());
}

#[test]
#[should_panic]
fn try_on_ndarray() {
    // Crashed (Invalid dimension: the shape of the index tensor should be the same as the value tensor: Index [1, 3] value [1, 10])
    scatter_notequal_forward_to_backward::<Autodiff<NdArray<f32>>>();
}

#[test]
#[should_panic]
fn try_on_candle() {
    // Crashed (called `Result::unwrap()` on an `Err` value: shape mismatch in scatter (indexes, src), lhs: [1, 3], rhs: [1, 10])
    scatter_notequal_forward_to_backward::<Autodiff<Candle<f32>>>();
}

#[test]
#[should_panic]
fn try_on_cuda() {
    // Crashed (called `Result::unwrap()` on an `Err` value: DriverError(CUDA_ERROR_ILLEGAL_ADDRESS, "an illegal memory access was encountered"))
    scatter_notequal_forward_to_backward::<Autodiff<Cuda<f32>>>();
}

// #[test]
// #[should_panic]
// fn try_on_cpu() {
//     // Crashes entire process (SIGSEGV: invalid memory reference)
//     scatter_notequal_forward_to_backward::<Autodiff<Cpu<f32>>>();
// }

#[test]
fn try_on_wgpu() {
    // OK
    scatter_notequal_forward_to_backward::<Autodiff<Wgpu<f32>>>();
}

#[test]
fn try_on_tch() {
    // OK
    scatter_notequal_forward_to_backward::<Autodiff<LibTorch<f32>>>();
}
```

**Comment by laggui** [comment]
Hey 👋 

As I mentioned in your linked PR (https://github.com/tracel-ai/burn/pull/4064#pullrequestreview-3504950517), burn's scatter is actually a scatter add (though we should extend that to support both more clearly).

The panic on backward shape mismatch is a bug though.

Thanks for flagging (and fixing!)
